### PR TITLE
Bump CoreDNS version to 1.6.5 and update manifest

### PR DESCRIPTION
**Pull request by rajansandeep**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

 - Bumps the CoreDNS version to 1.6.5
 - Updates the corefile-migration library to v1.0.4 which includes migration support up to CoreDNS v1.6.5

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
This PR is dependent on the CoreDNS 1.6.5 image be pushed to gcr.io for which https://github.com/kubernetes/kubernetes/issues/84993 has been opened.

/hold 

**Does this PR introduce a user-facing change?**:

```release-note
Kubeadm now includes CoreDNS version 1.6.5
 - `kubernetes` plugin adds metrics to measure kubernetes control plane latency.
 -  the `health` plugin now includes the `lameduck` option by default, which waits for a duration before shutting down.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
@rajansandeep this is very close the release, but i'm going to try to review it later.

> pull-kubernetes-e2e-kind — Job failed.
> ERROR: error building node image: command "docker save -o /tmp/kind-node-image755345787/bits/images/6.tar k8s.gcr.io/coredns:1.6.5" failed with error: exit status 1

we should push the image fist.
the kind job is now PR blocking.

**Comment by BenTheElder** [comment]
image is pushing https://github.com/kubernetes/kubernetes/issues/84993#issuecomment-552627578

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (BenTheElder) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go]: does kubeadm have an "addon manager" ? @neolit123

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: no, it does not have one, per se.
it has "phases" that manage addons.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: the comment can be:
`# Default replica count is 1`

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go]: is 12s the timeout for the health check in this case?
the timeout for CP components is 15s, so we may want to match that.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (BenTheElder) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: right, that's what I thought re: phases. 
I think the rest of the details in this comment make more sense for kube-up and less sense for kubeadm (presuming this is referring to the "Addon manager" in cluster/)

**Comment by BenTheElder** [comment]
/retest
kind passed

**Comment by rajansandeep** [comment]
/hold cancel 
Since the image seems to have been pushed to gcr

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/manifests.go]: @rajansandeep could you please explain the motivation?

my understanding is the following:
- we reduce the replica count to 1.
- coredns will deploy on the primary CP node (where kubeadm init is called).
- the anti-affinity rule makes sure that the Pod will not schedule on a Node that already has it.

if i'm not mistaken, this will not improve much what we have right now.
a problem we have currently, is that both replicas land on the same primary CP Node.

ideally what we want is a coredns instance to be deployed on all CP Nodes.
one way of doing that is with static-pods, but given we treat coredns as an addon we should use a DaemonSet with a NodeSelector that matches the kubeadm "master" node-role.

i'm going to experiment with that in a bit.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: sadly, by changing the coredns object type we are going to break a lot of users that have automation around `kubectl patch deployment coredns...`.

Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @neolit123 
With pod anti-affinity enabled and 2 coredns replicas:

- If a user has only a master node installed via `kubeadm init`, there will be one coredns pod in `running` state and one in `pending` state.
- The other coredns pod will remain in `pending` state and waits for scheduling until another worker node is created via `kubeadm join`.

Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: > a problem we have currently, is that both replicas land on the same primary CP Node.

Pod anti-affinity solves this problem.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: > With pod anti-affinity enabled and 2 coredns replicas:
> If a user has only a master node installed via kubeadm init, there will be one coredns pod in running state and one in pending state.
> The other coredns pod will remain in pending state and waits for scheduling until another worker node is created via kubeadm join.

this is true for 2 replicas, we don't want Pending pods because it will break e2e tests using our test suite, where all pods are expected to be ready.

>> a problem we have currently, is that both replicas land on the same primary CP Node.
> Pod anti-affinity solves this problem.

yes. but we reduce the replicas to 1, so if the primary CP node becomes NotReady (e.g. shutdown) the coredns service will go down too and the pod will not reschedule on a Ready node.

i guess i'm trying to see how 1 replica with anti-affinity is an improvement over 2 replicas without it.

like i've mentioned earlier, ideally we want a coredns DS for all CP nodes.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: if continuing to use a Deployment we might want to add these tolerations: https://github.com/kubernetes/kubernetes/issues/55713#issuecomment-538625309

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @chrisohaver PTAL too.

so basically i'm proposing that we keep the replica count to 2.
and introduce the following:
```
spec:
...
  tolerations:
  - key: "node.kubernetes.io/unreachable"
    operator: "Exists"
    effect: "NoExecute"
    tolerationSeconds: 15
  - key: "node.kubernetes.io/not-ready"
    operator: "Exists"
    effect: "NoExecute"
    tolerationSeconds: 15
```

this will improve the current deployment by rescheduling the coredns Pods if a Node becomes NotReady after 15 seconds.

i don't think the anti-affinity rule is needed here:
```
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: k8s-app
                operator: In
                values: ["kube-dns"]
            topologyKey: kubernetes.io/hostname
```

because with the current setup the deployment already does that.

Comment (rosti) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: Let's start with the more trivial questions here. Is this required for the CoreDNS version bump?
If so, why is this a patch release and not a minor version bump? If it's not required, can we split it and move it into a separate PR?

**Review by rosti — COMMENTED** [review]
Thanks @rajansandeep !

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: Yes, I've removed the pod anti-affinity changes from this PR and move it to another PR.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (chrisohaver) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: Actually - as I was describing the reasoning behind this, I realized that a timeout of 5 seconds should be all that is necessary.  When picking 12s, I was conflating the issue with the readiness/health check periods, which I don't think actually come into play.  The function of lameduck is to finish processing in flight queries before shutting down.  A lameduck of longer than 5 would typically be pointless, since most clients have a default timeout of 5 seconds (and thus would have stopped listening for a response after then).

**Comment by chrisohaver** [comment]
> Let's start with the more trivial questions here. Is this required for the CoreDNS version bump?

No it's not.

> If it's not required, can we split it and move it into a separate PR?

Yes - makes sense.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: yep, 5 seems good if that is sufficient.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: @rajansandeep do you agree with the change to 5 seconds?

**Comment by neolit123** [comment]
/approve
i think that there are a lot of flakes in CI right now.

File: cmd/kubeadm/app/phases/addons/dns/manifests.go
Comment (rajansandeep) [reply cmd/kubeadm/app/phases/addons/dns/manifests.go]: Yes, I agree. I'll push a commit to reflect those changes.

**Comment by neolit123** [comment]
/lgtm
thanks @rajansandeep

**Review by chrisohaver — APPROVED** [review]

**Comment by rajansandeep** [comment]
@neolit123 Does this need the milestone tag?

**Comment by neolit123** [comment]
> @neolit123 Does this need the milestone tag?

i don't think it does yet.

/retest

**Review by soltysh — APPROVED** [review]
/approve
dep updates

**Comment by rajansandeep** [comment]
/assign @liggitt 
For root approval for changes in vendor dep.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go
Comment (liggitt) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go]: is this a required change? will users with a custom dns config be broken if they don't make this change as well?

**Comment by liggitt** [comment]
/approve
vendor update looks good

/hold on the config compatibility question
kubeadm maintainers can unhold at will

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go
Comment (chrisohaver) [reply cmd/kubeadm/app/phases/addons/dns/dns_test.go]: It's not required.  It's just an improvement that reduces query failures during rolling upgrades.  The setting allows CoreDNS to complete in flight dns queries before exiting.

Without the setting, CoreDNS will not be broken.

**Comment by BenTheElder** [comment]
looks like I got scooped @neolit123 ... :prow_fire: :disappointed:

**Comment by neolit123** [comment]
> looks like I got scooped @neolit123 ... :prow_fire:

np

**Comment by neolit123** [comment]
canceling the hold as per @chrisohaver 's explanation here:
https://github.com/kubernetes/kubernetes/pull/85108#discussion_r345920135

thanks
/hold cancel

**Pull request merged**
